clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -364,6 +365,9 @@ private Cluster getClusterForCurrentTopics(Cluster cluster) {
             unauthorizedTopics.addAll(cluster.unauthorizedTopics());
             unauthorizedTopics.retainAll(this.topics.keySet());
 
+            invalidTopics.addAll(cluster.invalidTopics());
+            invalidTopics.addAll(this.cluster.invalidTopics());
```

Comment:
We could just use the same logic for `unauthorizedTopics` and allow these to expire as well. We could also bypass topic expiration and keep adding to `invalidTopics` as you have done, but in that case you can move the invalid topic check in `KafkaProducer.waitOnMetadata` (https://github.com/apache/kafka/pull/5247/files#diff-0436bd6994adccd181bfe6d02555cbc2R905) to the beginning and avoid adding it back to the topic set. I'm not sure at this point which option is better, but either is probably fine.

Reply:
If a topic is invalid, there is no logic to expiring the invalid topic names and adding the topic name back into the `Metadata.topic` map again, as the topic will very likely not become valid in the future.  I am going with expiring the invalid topic name one time (current workflow) and not adding the invalid topic back into `Metadata.topic` map.  I am moving the topic check in `KafkaProducer.waitOnMetadata` to the beginning of the function, as you suggested.